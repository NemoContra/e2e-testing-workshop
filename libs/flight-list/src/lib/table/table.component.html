<table class="table" *ngIf="flights" testId="table">
  <thead>
    <tr>
      <th>id</th>
      <th>from</th>
      <th>to</th>
      <th>date</th>
      <th>delayed</th>
    </tr>
  </thead>
  <tbody>
    <tr *ngFor="let flight of flights">
      <td>{{ flight.id }}</td>
      <td>{{ flight.from }}</td>
      <td>{{ flight.to }}</td>
      <td>{{ flight.date | date }}</td>
      <td>{{ flight.delayed ? 'yes' : 'no' }}</td>
    </tr>
  </tbody>
</table>
