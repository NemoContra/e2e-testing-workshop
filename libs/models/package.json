{
  "name": "@e2e-testing-workshop/models",
  "version": "0.0.1",
  "type": "commonjs"
}
