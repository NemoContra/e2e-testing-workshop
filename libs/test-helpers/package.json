{
  "name": "@e2e-testing-workshop/test-helpers",
  "version": "0.0.1",
  "type": "commonjs"
}
