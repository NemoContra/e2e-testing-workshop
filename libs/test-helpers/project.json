{
  "name": "test-helpers",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "libs/test-helpers/src",
  "projectType": "library",
  "targets": {
    "build": {
      "executor": "@nx/js:tsc",
      "outputs": ["{options.outputPath}"],
      "options": {
        "outputPath": "dist/libs/test-helpers",
        "main": "libs/test-helpers/src/index.ts",
        "tsConfig": "libs/test-helpers/tsconfig.lib.json",
        "assets": ["libs/test-helpers/*.md"]
      }
    },
    "lint": {
      "executor": "@nx/linter:eslint",
      "outputs": ["{options.outputFile}"],
      "options": {
        "lintFilePatterns": ["libs/test-helpers/**/*.ts"]
      }
    }
  },
  "tags": []
}
