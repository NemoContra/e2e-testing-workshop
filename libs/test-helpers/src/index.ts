export * from './lib/test-ids';
export * from './lib/mock-flights';
