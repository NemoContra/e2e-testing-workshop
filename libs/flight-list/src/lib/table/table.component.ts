import { ChangeDetectionStrategy, Component, Input } from '@angular/core';
import { DatePipe, NgForOf, NgIf } from '@angular/common';
import { Flight } from '@e2e-testing-workshop/models';
import { TestIdDirective } from '@e2e-testing-workshop/test-id';

@Component({
  selector: 'flight-table',
  standalone: true,
  imports: [DatePipe, NgForOf, NgIf, TestIdDirective],
  templateUrl: './table.component.html',
  styleUrls: ['./table.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class TableComponent {
  @Input() flights?: Flight[] | null;
}
