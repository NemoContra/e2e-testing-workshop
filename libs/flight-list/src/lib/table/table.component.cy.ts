import { MountConfig } from 'cypress/angular';
import { TableComponent } from './table.component';
import { mockFlights } from '@e2e-testing-workshop/test-helpers';
import { getTableCell, getTableHead } from '../../../cypress/support/app.po';

describe(TableComponent.name, () => {
  it('should test the table', () => {
    const mountConfig: MountConfig<TableComponent> = {
      componentProperties: {
        flights: mockFlights,
      },
    };
    cy.mount(TableComponent, mountConfig);

    getTableHead(0).contains('id');
    getTableHead(1).contains('from');
    getTableHead(2).contains('to');
    getTableHead(3).contains('date');
    getTableHead(4).contains('delayed');

    getTableCell(0, 0).contains('1');
    getTableCell(0, 1).contains('Graz');
    getTableCell(0, 2).contains('Mallorca');
    getTableCell(0, 3).contains('Oct 15, 2020');
    getTableCell(0, 4).contains('no');

    getTableCell(1, 0).contains('2');
    getTableCell(1, 1).contains('Hamburg');
    getTableCell(1, 2).contains('Mallorca');
    getTableCell(1, 3).contains('Oct 15, 2020');
    getTableCell(1, 4).contains('no');

    getTableCell(2, 0).contains('3');
    getTableCell(2, 1).contains('Lohr');
    getTableCell(2, 2).contains('Miami');
    getTableCell(2, 3).contains('Oct 15, 2020');
    getTableCell(2, 4).contains('no');
  });
});
