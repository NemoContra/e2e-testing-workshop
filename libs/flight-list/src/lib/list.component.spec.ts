import { ListComponent } from './list.component';
import { createHostFactory } from '@ngneat/spectator/jest';
import { TableComponent } from './table/table.component';
import { MockComponent } from 'ng-mocks';
import { FlightService } from '@e2e-testing-workshop/services';
import { ReplaySubject } from 'rxjs';
import { Flight } from '@e2e-testing-workshop/models';
import { mockFlights } from '@e2e-testing-workshop/test-helpers';

describe('ListComponent', () => {
  const flights$$ = new ReplaySubject<Flight[]>(1);
  const createComponent = createHostFactory({
    component: ListComponent,
    providers: [
      {
        provide: FlightService,
        useFactory: () => ({
          getFlights: jest.fn().mockReturnValue(flights$$),
        }),
      },
    ],
    overrideComponents: [
      [
        TableComponent,
        {
          remove: { imports: [TableComponent] },
          add: { imports: [MockComponent(TableComponent)] },
        },
      ],
    ],
  });

  it('render an empty table', () => {
    const spectator = createComponent('<flight-list></flight-list>');
    expect(spectator.fixture).toMatchSnapshot();
    expect(spectator.query(TableComponent)?.flights).toBeNull();
  });

  it('should resolve the data from http and render the table', () => {
    const spectator = createComponent('<flight-list></flight-list>');
    flights$$.next(mockFlights);
    spectator.detectChanges();
    expect(spectator.fixture).toMatchSnapshot();
    expect(spectator.query(TableComponent)?.flights).toEqual(mockFlights);
  });
});
