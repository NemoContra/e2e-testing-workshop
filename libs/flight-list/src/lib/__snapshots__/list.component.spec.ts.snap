// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ListComponent render an empty table 1`] = `
<ng-component>
  <flight-list>
    <flight-table>
    </flight-table>
  </flight-list>
</ng-component>
`;

exports[`ListComponent should resolve the data from http and render the table 1`] = `
<ng-component>
  <flight-list>
    <flight-table>
      <table
        class="table"
        data-testid="table"
        testid="table"
      >
        <thead>
          <tr>
            <th>
              id
            </th>
            <th>
              from
            </th>
            <th>
              to
            </th>
            <th>
              date
            </th>
            <th>
              delayed
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              1
            </td>
            <td>
              Graz
            </td>
            <td>
              Mallorca
            </td>
            <td>
              Oct 15, 2020
            </td>
            <td>
              no
            </td>
          </tr>
          <tr>
            <td>
              2
            </td>
            <td>
              Hamburg
            </td>
            <td>
              Mallorca
            </td>
            <td>
              Oct 15, 2020
            </td>
            <td>
              no
            </td>
          </tr>
          <tr>
            <td>
              3
            </td>
            <td>
              Lohr
            </td>
            <td>
              Miami
            </td>
            <td>
              Oct 15, 2020
            </td>
            <td>
              no
            </td>
          </tr>
        </tbody>
      </table>
    </flight-table>
  </flight-list>
</ng-component>
`;
